File: 13-Decorator/src/cpp/main.cpp
#include <iostream>
#include <iomanip>
#include <memory>
#include "Espresso.h"
#include "Latte.h"
#include "MilkDecorator.h"
#include "WhipDecorator.h"
#include "CaramelDecorator.h"

void print_coffee(const std::string& label, const Coffee& coffee) {
    std::cout << "  " << std::left << std::setw(15) << (label + ":")
              << coffee.get_description() << " — $"
              << std::fixed << std::setprecision(2) << coffee.get_cost() << "\n";
}

int main() {
    std::cout << "=== Decorator Pattern Demo (C++) ===\n\n";

    // 1. Plain espresso
    auto plain = std::make_unique<Espresso>();
    print_coffee("Plain", *plain);

    // 2. Espresso + Milk
    auto with_milk = std::make_unique<MilkDecorator>(std::make_unique<Espresso>());
    print_coffee("With milk", *with_milk);

    // 3. Stacked: Espresso + Milk + Whip + Caramel
    auto loaded = std::make_unique<CaramelDecorator>(
                    std::make_unique<WhipDecorator>(
                      std::make_unique<MilkDecorator>(
                        std::make_unique<Espresso>())));
    print_coffee("Loaded", *loaded);

    // 4. Latte + Whip
    auto latte_whip = std::make_unique<WhipDecorator>(std::make_unique<Latte>());
    print_coffee("Latte+Whip", *latte_whip);

    // 5. Double milk
    auto double_milk = std::make_unique<MilkDecorator>(
                         std::make_unique<MilkDecorator>(
                           std::make_unique<Espresso>()));
    print_coffee("Double milk", *double_milk);

    return 0;
}

File: 13-Decorator/src/rust/main.rs
mod coffee;
mod base_coffees;
mod decorators;

use coffee::Coffee;
use base_coffees::{Espresso, Latte};
use decorators::{MilkDecorator, WhipDecorator, CaramelDecorator};

fn print_coffee(label: &str, coffee: &dyn Coffee) {
    println!("  {:<15} {} — ${:.2}", format!("{}:", label), coffee.get_description(), coffee.get_cost());
}

fn main() {
    println!("=== Decorator Pattern Demo (Rust) ===\n");

    // 1. Plain espresso
    let plain = Espresso;
    print_coffee("Plain", &plain);

    // 2. Espresso + Milk
    let with_milk = MilkDecorator::new(Box::new(Espresso));
    print_coffee("With milk", &with_milk);

    // 3. Stacked: Espresso + Milk + Whip + Caramel
    let loaded = CaramelDecorator::new(
        Box::new(WhipDecorator::new(
            Box::new(MilkDecorator::new(
                Box::new(Espresso))))));
    print_coffee("Loaded", &loaded);

    // 4. Latte + Whip
    let latte_whip = WhipDecorator::new(Box::new(Latte));
    print_coffee("Latte+Whip", &latte_whip);

    // 5. Double milk
    let double_milk = MilkDecorator::new(
        Box::new(MilkDecorator::new(
            Box::new(Espresso))));
    print_coffee("Double milk", &double_milk);
}
